### PR TITLE
MODINREACH-45 Create Backend API Endpoint to CRUD a Mapping of FOLIO Material Types to INN-Reach Central Item Types

### PR DESCRIPTION
**Pull request by dmtkachenko**
## Purpose
Create an API endpoint that allows CRUD operations on data model defined for Folio material type mapping to INN-Reach central item types.
JIRA: [MODINREACH-45](https://issues.folio.org/browse/MODINREACH-45)

## Approach
* describe API endpoints in OpenAPI spec
* introduce `MaterialTypeMappingService` service to handle the required business operations
* implement controller to respond to CRUD operation requests from UI 
* add additional exceptions handlers to correctly manage validation exceptions and data integrity constraint violations

## Learning
* to control how `openapi-generator-maven-plugin` combines API methods into a controller interface `useTags` config option can be used together with `tags` element in OpenAPI spec. See [inn-reach.yaml ](https://github.com/folio-org/mod-inn-reach/pull/20/files#diff-fc664aec3b7a5db198553c50d0ada5557750696120f92d57865736d54bffd1ee) and [pom.xml](https://github.com/folio-org/mod-inn-reach/pull/20/files#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8)

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/resources/swagger.api/schemas/innReachLocationsDTO.json [hunk 1]
```diff
@@ -7,6 +7,7 @@
       "type": "integer"
     },
     "locations": {
+      "description": "List of INN-Reach locations",
```
Comment (dcrossleyau) [inline src/main/resources/swagger.api/schemas/innReachLocationsDTO.json:10]: Hooray. Thanks.

File: src/main/java/org/folio/innreach/domain/service/impl/MaterialTypeMappingServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  private final MaterialTypeMappingRepository repository;
+  private final MaterialTypeMappingMapper mapper;
+
+  
+  @Override
+  @Transactional(readOnly = true)
+  public MaterialTypeMappingsDTO getAllMappings(UUID centralServerId, Integer offset, Integer limit) {
+    var example = mappingExampleWithServerId(centralServerId);
+
+    var totalCount = repository.count(example);
```
Comment (andrew-khodyka) [inline src/main/java/org/folio/innreach/domain/service/impl/MaterialTypeMappingServiceImpl.java]: Do we really need this additional count query?

File: src/main/java/org/folio/innreach/domain/service/impl/MaterialTypeMappingServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+  private static Example<MaterialTypeMapping> mappingExampleWithServerIdAndId(UUID centralServerId, UUID id) {
+    var toFind = new MaterialTypeMapping();
+    toFind.setId(id);
+    toFind.setCentralServer(centralServerRef(centralServerId));
+
+    return Example.of(toFind);
+  }
+
+  private static CentralServer centralServerRef(UUID centralServerId) {
```
Comment (andrew-khodyka) [inline src/main/java/org/folio/innreach/domain/service/impl/MaterialTypeMappingServiceImpl.java:105]: There's the method to create a reference without querying the database
**repository.getOne(id)**

File: src/main/java/org/folio/innreach/domain/service/impl/MaterialTypeMappingServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+  }
+
+  private static Example<MaterialTypeMapping> mappingExampleWithServerId(UUID centralServerId) {
+    var toFind = new MaterialTypeMapping();
+    toFind.setCentralServer(centralServerRef(centralServerId));
+
+    return Example.of(toFind);
+  }
+
+  private static Example<MaterialTypeMapping> mappingExampleWithServerIdAndId(UUID centralServerId, UUID id) {
```
Comment (andrew-khodyka) [inline src/main/java/org/folio/innreach/domain/service/impl/MaterialTypeMappingServiceImpl.java:97]: Does this method have to be static?

File: src/main/java/org/folio/innreach/domain/service/impl/MaterialTypeMappingServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    repository.delete(mapping);
+  }
+
+  private MaterialTypeMapping findMapping(UUID centralServerId, UUID id) {
+    return repository.findOne(mappingExampleWithServerIdAndId(centralServerId, id))
+        .orElseThrow(() -> new EntityNotFoundException("Material type mapping not found: id = " + id +
+            ", centralServerId = " + centralServerId));
+  }
+
+  private static Example<MaterialTypeMapping> mappingExampleWithServerId(UUID centralServerId) {
```
Comment (andrew-khodyka) [inline src/main/java/org/folio/innreach/domain/service/impl/MaterialTypeMappingServiceImpl.java:90]: Does this method have to be static?

File: src/main/java/org/folio/innreach/domain/service/impl/MaterialTypeMappingServiceImpl.java (hunk 2 above)
Comment (dmtkachenko) [reply src/main/java/org/folio/innreach/domain/service/impl/MaterialTypeMappingServiceImpl.java]: Yes, we do.. It calculates the total number of records not limited by `limit` and `offset` parameters. It's a common agreement (contract) in Folio to return the total count for such kind of queries. 
I know it's settings and in most cases query will retrieve everything, but still

File: src/main/java/org/folio/innreach/domain/service/impl/MaterialTypeMappingServiceImpl.java (hunk 3 above)
Comment (dmtkachenko) [reply src/main/java/org/folio/innreach/domain/service/impl/MaterialTypeMappingServiceImpl.java:105]: Yes thanks, I'm aware of this method. But in this case I don't need a proxied hibernate reference but a simple pojo (entity) with the only UUID field initialized. 
Agree that `getOne()` will work too

File: src/main/java/org/folio/innreach/domain/service/impl/MaterialTypeMappingServiceImpl.java (hunk 4 above)
Comment (dmtkachenko) [reply src/main/java/org/folio/innreach/domain/service/impl/MaterialTypeMappingServiceImpl.java:97]: I tend to create utility methods as static. It has no relation with the object state, so can be "detached" from it. Also it has better readability for me in the code that uses it
So it's just my preference

File: src/main/java/org/folio/innreach/domain/service/impl/MaterialTypeMappingServiceImpl.java (hunk 5 above)
Comment (dmtkachenko) [reply src/main/java/org/folio/innreach/domain/service/impl/MaterialTypeMappingServiceImpl.java:90]: See my answer in other similar post

File: src/main/java/org/folio/innreach/domain/service/impl/MaterialTypeMappingServiceImpl.java (hunk 2 above)
Comment (andrew-khodyka) [reply src/main/java/org/folio/innreach/domain/service/impl/MaterialTypeMappingServiceImpl.java]: You don't need to query for it explicitly.
Page object can return the total number of elements in the database, 
Please, check **mappings.getTotalElements()**

Comment (dmtkachenko) [reply src/main/java/org/folio/innreach/domain/service/impl/MaterialTypeMappingServiceImpl.java]: Thanks for the tip. Will take a look at the method

**Review by andrew-khodyka — APPROVED** [review]

**Pull request merged**